### PR TITLE
Allow more than 10 assignments on Mechanical Turk

### PR DESCRIPTION
**Pull request by MatthewWilkes**
## Description
This creates a recruiter that can get more than 10 participants on mechanical turk, and a chain implementation that handles the initial glut of participants. It also adds a bot for MCMCP to assist with testing this.

## Motivation and Context
Mechanical turk's pricing increases if you need more than 10 assignments, so if your initial recruitment size is less than 10 it will fail on extending past that limit. This pre-allocates 10 participants and introduces a chain and a recruiter type to make it compatible with experiments without them needing to specifically code for the mturk limitations.

## How Has This Been Tested?
Unit tests only, at this stage.

### CODE REVIEW

**Comment by MatthewWilkes** [comment]
I'm not sure how coverage has decreased, as I wrote additional tests, but leave this with me @suchow and I'll fix it up before you need to look at it.

**Comment by suchow** [comment]
@MatthewWilkes Pretty sure @codecov[bot] is thoroughly confused, it's been doing this for the past few days.

**Comment by MatthewWilkes** [comment]
Hi @suchow 

Your understanding of the idea for the large recruiter is right. The idea is that the responsibility for handling the mturk 10 HIT requirement shouldn't be on the experiment, it should be possible to switch between different recruiters without having a difference on the parameters of the experiment.

I think there are effectively two parts to this that we should treat independently:

1. MTurk sometimes requires you to recruit more participants than you need, and that shouldn't be something the experiment needs to worry about
2. Sometimes a network gets participants before it's ready, and that's not something the experiment should care about either.

There are other times that issue 2 could happen, like using Facebook adverts to recruit people, as you have no control over when people will click the advert. Also, other services might have similar problems to 1 with different subtleties, so requiring changes to experiment parameters would make reproducability tough.

As to your two questions, I've made some changes this morning.

1. You're absolutely right about state. What I've done is moved the state to the redis queue, but I almost think we should support recruiter state in the DB. What do you think?
2. Don't forget that all 10 will finish the recruitment. If a participant doesn't reach the end state then the error handling will recruit another participant directly from MTurk. When the first one completes successfully there won't be another participant recruited, but when the 10th completes there will be an 11th recruited. This way the experiment tries to recruit as normal, but nothing happens until the initial glut of users has been used up.

That way, regardless of the number of people that had to be recruited for operational reasons, once the experiment gets going it will behave in the same way as if it was running on an recruiter that behaved normally. In the previous implementation, you'd be forced to have 10 concurrent HITs, as replacements would be recruited for everyone who completed. That'd prevent you from using MTurk for large experiments that want fewer than 10 concurrent users.

Does that make sense?

File: demos/bartlett1932/experiment.py [hunk 1]
```diff
@@ -50,7 +53,7 @@ def setup(self):
 
     def create_network(self):
         """Return a new network."""
-        return Chain(max_size=5)
+        return DelayedChain()
```
Comment (suchow) [inline demos/bartlett1932/experiment.py]: I think removing `max_size` here will make this demo run indefinitely?

**Comment by MatthewWilkes** [comment]
That error was some code that had survived from a previous implementation, it can be safely removed (and I've done so).

**Comment by suchow** [comment]
@MatthewWilkes I spoke just a bit too soon. This appears to be feature-complete, functional, and tested, but I see a side effect that I can't seem to fix. When I run the MCMCP demo, I'm seeing the following "Connection aborted" error. This happens only for the MCMCP demo, and not for Bartett1932, which is what is making me think that it has something to do with this PR. The traceback is:

```
(dallinger) ➜  mcmcp git:(long-chain-alt) ✗ dallinger debug --verbose

    ____        ____
   / __ \____ _/ / (_)___  ____ ____  _____
  / / / / __ `/ / / / __ \/ __ `/ _ \/ ___/
 / /_/ / /_/ / / / / / / / /_/ /  __/ /
/_____/\__,_/_/_/_/_/ /_/\__, /\___/_/
                        /____/
                                   v3.4.0

                Laboratory automation for
       the behavioral and social sciences.


❯❯ Experiment id is 63c291bd-98bf-42ec-b6ea-ed0be834bb06
/var/folders/l4/pjvjhj2n1qg_2s0t23v2l4h40000gn/T/tmpocLca2/63c291bd-98bf-42ec-b6ea-ed0be834bb06

❯❯ Starting up the server...
[WARN] No ENV file found
12:51:55 PM worker.1 |  12:51:55 RQ worker u'rq:worker:Jordans-MacBook-Pro.81822' started, version 0.8.1
12:51:55 PM worker.1 |  12:51:55 Cleaning registries for queue: high
12:51:55 PM worker.1 |  12:51:55 Cleaning registries for queue: default
12:51:55 PM worker.1 |  12:51:55 Cleaning registries for queue: low
12:51:55 PM worker.1 |  12:51:55
12:51:55 PM worker.1 |  12:51:55 *** Listening on high, default, low...
12:51:57 PM web.2   |  [2017-08-21 12:51:57 -0400] [81821] [DEBUG] Current configuration:
12:51:57 PM web.2   |    proxy_protocol: False
12:51:57 PM web.2   |    worker_connections: 1000
12:51:57 PM web.2   |    statsd_host: None
12:51:57 PM web.2   |    max_requests_jitter: 0
12:51:57 PM web.2   |    post_fork: <function post_fork at 0x107bcbed8>
12:51:57 PM web.2   |    errorlog: -
12:51:57 PM web.2   |    enable_stdio_inheritance: False
12:51:57 PM web.2   |    worker_class: geventwebsocket.gunicorn.workers.GeventWebSocketWorker
12:51:57 PM web.2   |    ssl_version: 2
12:51:57 PM web.2   |    suppress_ragged_eofs: True
12:51:57 PM web.2   |    syslog: False
12:51:57 PM web.2   |    syslog_facility: user
12:51:57 PM web.2   |    when_ready: <function when_ready at 0x107bdc938>
12:51:57 PM web.2   |    pre_fork: <function pre_fork at 0x107bcbd70>
12:51:57 PM web.2   |    cert_reqs: 0
12:51:57 PM web.2   |    preload_app: False
12:51:57 PM web.2   |    keepalive: 2
12:51:57 PM web.2   |    accesslog: -
12:51:57 PM web.2   |    group: 20
12:51:57 PM web.2   |    graceful_timeout: 30
12:51:57 PM web.2   |    do_handshake_on_connect: False
12:51:57 PM web.2   |    spew: False
12:51:57 PM web.2   |    workers: 9
12:51:57 PM web.2   |    proc_name: dallinger_experiment_server
12:51:57 PM web.2   |    sendfile: None
12:51:57 PM web.2   |    pidfile: None
12:51:57 PM web.2   |    umask: 0
12:51:57 PM web.2   |    on_reload: <function on_reload at 0x107bcbaa0>
12:51:57 PM web.2   |    pre_exec: <function pre_exec at 0x107bd6500>
12:51:57 PM web.2   |    worker_tmp_dir: None
12:51:57 PM web.2   |    limit_request_fields: 100
12:51:57 PM web.2   |    pythonpath: None
12:51:57 PM web.2   |    on_exit: <function on_exit at 0x107bd6cf8>
12:51:57 PM web.2   |    config: None
12:51:57 PM web.2   |    logconfig: None
12:51:57 PM web.2   |    check_config: False
12:51:57 PM web.2   |    statsd_prefix:
12:51:57 PM web.2   |    secure_scheme_headers: {'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}
12:51:57 PM web.2   |    reload_engine: auto
12:51:57 PM web.2   |    proxy_allow_ips: ['127.0.0.1']
12:51:57 PM web.2   |    pre_request: <function pre_request at 0x107bd6668>
12:51:57 PM web.2   |    post_request: <function post_request at 0x107bd6758>
12:51:57 PM web.2   |    forwarded_allow_ips: ['127.0.0.1']
12:51:57 PM web.2   |    worker_int: <function worker_int at 0x107bd6230>
12:51:57 PM web.2   |    raw_paste_global_conf: []
12:51:57 PM web.2   |    threads: 1
12:51:57 PM web.2   |    max_requests: 0
12:51:57 PM web.2   |    chdir: /private/var/folders/l4/pjvjhj2n1qg_2s0t23v2l4h40000gn/T/tmpocLca2/63c291bd-98bf-42ec-b6ea-ed0be834bb06
12:51:57 PM web.2   |    daemon: False
12:51:57 PM web.2   |    user: 501
12:51:57 PM web.2   |    limit_request_line: 0
12:51:57 PM web.2   |    access_log_format: %(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"
12:51:57 PM web.2   |    certfile: None
12:51:57 PM web.2   |    on_starting: <function on_starting at 0x107bcb938>
12:51:57 PM web.2   |    post_worker_init: <function post_worker_init at 0x107bd60c8>
12:51:57 PM web.2   |    child_exit: <function child_exit at 0x107bd68c0>
12:51:57 PM web.2   |    worker_exit: <function worker_exit at 0x107bd6a28>
12:51:57 PM web.2   |    paste: None
12:51:57 PM web.2   |    default_proc_name: gunicorn
12:51:57 PM web.2   |    syslog_addr: unix:///var/run/syslog
12:51:57 PM web.2   |    syslog_prefix: None
12:51:57 PM web.2   |    ciphers: TLSv1
12:51:57 PM web.2   |    worker_abort: <function worker_abort at 0x107bd6398>
12:51:57 PM web.2   |    loglevel: debug
12:51:57 PM web.2   |    bind: ['0.0.0.0:5001']
12:51:57 PM web.2   |    raw_env: []
12:51:57 PM web.2   |    initgroups: False
12:51:57 PM web.2   |    capture_output: False
12:51:57 PM web.2   |    reload: False
12:51:57 PM web.2   |    limit_request_field_size: 8190
12:51:57 PM web.2   |    nworkers_changed: <function nworkers_changed at 0x107bd6b90>
12:51:57 PM web.2   |    timeout: 30
12:51:57 PM web.2   |    keyfile: None
12:51:57 PM web.2   |    ca_certs: None
12:51:57 PM web.2   |    tmp_upload_dir: None
12:51:57 PM web.2   |    backlog: 2048
12:51:57 PM web.2   |    logger_class: gunicorn.glogging.Logger
12:51:57 PM web.2   |  [2017-08-21 12:51:57 -0400] [81821] [INFO] Starting gunicorn 19.7.1
12:51:57 PM web.2   |  [2017-08-21 12:51:57 -0400] [81821] [ERROR] Connection in use: ('0.0.0.0', 5001)
12:51:57 PM web.2   |  [2017-08-21 12:51:57 -0400] [81821] [DEBUG] connection to ('0.0.0.0', 5001) failed: [Errno 48] Address already in use
12:51:57 PM web.2   |  [2017-08-21 12:51:57 -0400] [81821] [ERROR] Retrying in 1 second.
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81820] [DEBUG] Current configuration:
12:51:57 PM web.1   |    proxy_protocol: False
12:51:57 PM web.1   |    worker_connections: 1000
12:51:57 PM web.1   |    statsd_host: None
12:51:57 PM web.1   |    max_requests_jitter: 0
12:51:57 PM web.1   |    post_fork: <function post_fork at 0x1125aced8>
12:51:57 PM web.1   |    errorlog: -
12:51:57 PM web.1   |    enable_stdio_inheritance: False
12:51:57 PM web.1   |    worker_class: geventwebsocket.gunicorn.workers.GeventWebSocketWorker
12:51:57 PM web.1   |    ssl_version: 2
12:51:57 PM web.1   |    suppress_ragged_eofs: True
12:51:57 PM web.1   |    syslog: False
12:51:57 PM web.1   |    syslog_facility: user
12:51:57 PM web.1   |    when_ready: <function when_ready at 0x1125bd938>
12:51:57 PM web.1   |    pre_fork: <function pre_fork at 0x1125acd70>
12:51:57 PM web.1   |    cert_reqs: 0
12:51:57 PM web.1   |    preload_app: False
12:51:57 PM web.1   |    keepalive: 2
12:51:57 PM web.1   |    accesslog: -
12:51:57 PM web.1   |    group: 20
12:51:57 PM web.1   |    graceful_timeout: 30
12:51:57 PM web.1   |    do_handshake_on_connect: False
12:51:57 PM web.1   |    spew: False
12:51:57 PM web.1   |    workers: 9
12:51:57 PM web.1   |    proc_name: dallinger_experiment_server
12:51:57 PM web.1   |    sendfile: None
12:51:57 PM web.1   |    pidfile: None
12:51:57 PM web.1   |    umask: 0
12:51:57 PM web.1   |    on_reload: <function on_reload at 0x1125acaa0>
12:51:57 PM web.1   |    pre_exec: <function pre_exec at 0x1125b7500>
12:51:57 PM web.1   |    worker_tmp_dir: None
12:51:57 PM web.1   |    limit_request_fields: 100
12:51:57 PM web.1   |    pythonpath: None
12:51:57 PM web.1   |    on_exit: <function on_exit at 0x1125b7cf8>
12:51:57 PM web.1   |    config: None
12:51:57 PM web.1   |    logconfig: None
12:51:57 PM web.1   |    check_config: False
12:51:57 PM web.1   |    statsd_prefix:
12:51:57 PM web.1   |    secure_scheme_headers: {'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}
12:51:57 PM web.1   |    reload_engine: auto
12:51:57 PM web.1   |    proxy_allow_ips: ['127.0.0.1']
12:51:57 PM web.1   |    pre_request: <function pre_request at 0x1125b7668>
12:51:57 PM web.1   |    post_request: <function post_request at 0x1125b7758>
12:51:57 PM web.1   |    forwarded_allow_ips: ['127.0.0.1']
12:51:57 PM web.1   |    worker_int: <function worker_int at 0x1125b7230>
12:51:57 PM web.1   |    raw_paste_global_conf: []
12:51:57 PM web.1   |    threads: 1
12:51:57 PM web.1   |    max_requests: 0
12:51:57 PM web.1   |    chdir: /private/var/folders/l4/pjvjhj2n1qg_2s0t23v2l4h40000gn/T/tmpocLca2/63c291bd-98bf-42ec-b6ea-ed0be834bb06
12:51:57 PM web.1   |    daemon: False
12:51:57 PM web.1   |    user: 501
12:51:57 PM web.1   |    limit_request_line: 0
12:51:57 PM web.1   |    access_log_format: %(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"
12:51:57 PM web.1   |    certfile: None
12:51:57 PM web.1   |    on_starting: <function on_starting at 0x1125ac938>
12:51:57 PM web.1   |    post_worker_init: <function post_worker_init at 0x1125b70c8>
12:51:57 PM web.1   |    child_exit: <function child_exit at 0x1125b78c0>
12:51:57 PM web.1   |    worker_exit: <function worker_exit at 0x1125b7a28>
12:51:57 PM web.1   |    paste: None
12:51:57 PM web.1   |    default_proc_name: gunicorn
12:51:57 PM web.1   |    syslog_addr: unix:///var/run/syslog
12:51:57 PM web.1   |    syslog_prefix: None
12:51:57 PM web.1   |    ciphers: TLSv1
12:51:57 PM web.1   |    worker_abort: <function worker_abort at 0x1125b7398>
12:51:57 PM web.1   |    loglevel: debug
12:51:57 PM web.1   |    bind: ['0.0.0.0:5000']
12:51:57 PM web.1   |    raw_env: []
12:51:57 PM web.1   |    initgroups: False
12:51:57 PM web.1   |    capture_output: False
12:51:57 PM web.1   |    reload: False
12:51:57 PM web.1   |    limit_request_field_size: 8190
12:51:57 PM web.1   |    nworkers_changed: <function nworkers_changed at 0x1125b7b90>
12:51:57 PM web.1   |    timeout: 30
12:51:57 PM web.1   |    keyfile: None
12:51:57 PM web.1   |    ca_certs: None
12:51:57 PM web.1   |    tmp_upload_dir: None
12:51:57 PM web.1   |    backlog: 2048
12:51:57 PM web.1   |    logger_class: gunicorn.glogging.Logger
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81820] [INFO] Starting gunicorn 19.7.1
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81820] [DEBUG] Arbiter booted
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81820] [INFO] Listening at: http://0.0.0.0:5000 (81820)
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81820] [INFO] Using worker: geventwebsocket.gunicorn.workers.GeventWebSocketWorker
12:51:57 PM web.1   |  2017-08-21 12:51:57,576 Ready.
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81889] [INFO] Booting worker with pid: 81889
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81890] [INFO] Booting worker with pid: 81890
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81891] [INFO] Booting worker with pid: 81891
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81893] [INFO] Booting worker with pid: 81893
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81894] [INFO] Booting worker with pid: 81894
12:51:57 PM web.1   |  [2017-08-21 12:51:57 -0400] [81895] [INFO] Booting worker with pid: 81895
12:51:58 PM web.1   |  [2017-08-21 12:51:58 -0400] [81896] [INFO] Booting worker with pid: 81896
12:51:58 PM web.1   |  [2017-08-21 12:51:58 -0400] [81897] [INFO] Booting worker with pid: 81897
12:51:58 PM web.1   |  [2017-08-21 12:51:58 -0400] [81898] [INFO] Booting worker with pid: 81898
12:51:58 PM web.1   |  [2017-08-21 12:51:58 -0400] [81820] [DEBUG] 9 workers

❯❯ Server is running on http://0.0.0.0:5000. Press Ctrl+C to exit.

❯❯ Launching the experiment...

❯❯ Completed debugging of experiment with id 63c291bd-98bf-42ec-b6ea-ed0be834bb06

❯❯ Cleaning up local Heroku process...

❯❯ Local Heroku was already terminated.

❯❯ Traceback (most recent call last):
  File "/Users/jordansuchow/Dropbox/Berkeley/Projects/Current/Dallinger/dallinger/heroku/tools.py", line 169, in stop
    os.killpg(os.getpgid(self._process.pid), signal)
OSError: [Errno 3] No such process

Traceback (most recent call last):
  File "/usr/local/bin/dallinger", line 11, in <module>
    load_entry_point('dallinger', 'console_scripts', 'dallinger')()
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/jordansuchow/Dropbox/Berkeley/Projects/Current/Dallinger/dallinger/command_line.py", line 409, in debug
    debugger.run()
  File "/Users/jordansuchow/Dropbox/Berkeley/Projects/Current/Dallinger/dallinger/command_line.py", line 819, in run
    self.execute(wrapper)
  File "/Users/jordansuchow/Dropbox/Berkeley/Projects/Current/Dallinger/dallinger/command_line.py", line 865, in execute
    _handle_launch_data('{}/launch'.format(base_url), error=self.out.error)
  File "/Users/jordansuchow/Dropbox/Berkeley/Projects/Current/Dallinger/dallinger/command_line.py", line 384, in _handle_launch_data
    launch_request = requests.post(url)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 112, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/adapters.py", line 490, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', error(54, 'Connection reset by peer'))
```

**Pull request merged**